File: services/pipeline-service/app/api/v1/routes/demands.py
"""Criacao e consulta filtrada de negociacoes (RF02, RF03).

Erro nenhum e traduzido aqui: o servico levanta ``PlatformError`` e os handlers
registrados em ``app.core.exceptions`` devolvem o envelope unico da plataforma
(RNF02). Antes existia uma traducao local nesta rota, e o servico respondia
``422`` em dois formatos conforme o caminho.
"""

from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.exceptions import ValidationError, issue_de_validacao
from app.db.session import get_session
from app.domain.enums import DemandStatus, PipelineStage
from app.repositories.demand_repository import DemandRepository
from app.schemas.common import PaginatedResponse
from app.schemas.demand import (
    DemandCreate,
    DemandDetail,
    DemandRead,
    DemandUpdate,
)
from app.schemas.error import ErrorResponse
from app.services.demand_service import DemandService

router = APIRouter(prefix="/demands", tags=["demands"])


@router.get(
    "",
    response_model=PaginatedResponse[DemandRead],
    summary="Listar e filtrar demandas",
    description=(
        "Lista demandas por cliente, etapa, responsavel, periodo de criacao e situacao. "
        "Os filtros informados sao combinados com AND."
    ),
    responses={
        422: {"model": ErrorResponse, "description": "Filtros invalidos."},
    },
)
def list_demands(
    session: Annotated[Session, Depends(get_session)],
    client_id: UUID | None = None,
    stage: PipelineStage | None = None,
    owner_id: UUID | None = None,
    created_from: Annotated[
        datetime | None,
        Query(alias="from", description="Inicio inclusivo do periodo de criacao, com fuso."),
    ] = None,
    created_to: Annotated[
        datetime | None,
        Query(alias="to", description="Fim inclusivo do periodo de criacao, com fuso."),
    ] = None,
    demand_status: Annotated[
        DemandStatus | None,
        Query(alias="status", description="Situacao atual da demanda."),
    ] = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> PaginatedResponse[DemandRead]:
    """Aplica filtros combinaveis e devolve o total antes da paginacao."""

    _validate_period(created_from, created_to)

    items, total = DemandRepository(session).list(
        client_id=client_id,
        stage=stage,
        owner_id=owner_id,
        created_from=created_from,
        created_to=created_to,
        status=demand_status,
        limit=limit,
        offset=offset,
    )
    return PaginatedResponse[DemandRead](
        items=[DemandRead.model_validate(item) for item in items],
        total=total,
        page=(offset // limit) + 1,
        size=limit,
    )


def _validate_period(created_from: datetime | None, created_to: datetime | None) -> None:
    """Exige fuso e uma faixa cronologica coerente para o filtro de periodo.

    Levanta ``ValidationError`` para que a resposta saia pelo mesmo handler do
    restante do servico, com o ``details.issues`` de sempre.
    """

    issues: list[dict[str, object]] = []
    if created_from is not None and created_from.utcoffset() is None:
        issues.append(
            issue_de_validacao(
                localizacao=["query", "from"],
                mensagem="A data inicial deve informar o fuso horario.",
                tipo="value_error.timezone",
            )
        )
    if created_to is not None and created_to.utcoffset() is None:
        issues.append(
            issue_de_validacao(
                localizacao=["query", "to"],
                mensagem="A data final deve informar o fuso horario.",
                tipo="value_error.timezone",
            )
        )
    if (
        not issues
        and created_from is not None
        and created_to is not None
        and created_from > created_to
    ):
        issues.append(
            issue_de_validacao(
                localizacao=["query", "to"],
                mensagem="A data final deve ser maior ou igual a data inicial.",
                tipo="value_error.period",
            )
        )

    if issues:
        raise ValidationError(
            code="VALIDATION_ERROR",
            message="Os dados informados sao invalidos.",
            details={"issues": issues},
        )


@router.post(
    "",
    response_model=DemandRead,
    status_code=status.HTTP_201_CREATED,
    summary="Criar demanda vinculada a cliente",
    description=(
        "Cria uma negociacao com cliente obrigatorio, na situacao ABERTA e etapa CAPTACAO. "
        "description registra o contexto; owner_id indica um usuario responsavel existente."
    ),
    responses={
        404: {"model": ErrorResponse, "description": "Cliente ou responsavel inexistente."},
        409: {
            "model": ErrorResponse,
            "description": "Referencia removida durante a criacao.",
        },
        422: {"model": ErrorResponse, "description": "Dados de entrada invalidos."},
    },
)
def create_demand(
    data: DemandCreate,
    session: Annotated[Session, Depends(get_session)],
) -> DemandRead:
    """Aplica o caso de uso e traduz apenas erros previsiveis do dominio."""

    return DemandRead.model_validate(DemandService(session).create(data))


@router.get(
    "/{demand_id}",
    response_model=DemandDetail,
    summary="Consultar contexto da demanda",
    description=(
        "Retorna o contexto da demanda, o cliente, a etapa corrente e os artefatos "
        "vinculados com suas versoes."
    ),
    responses={
        404: {"model": ErrorResponse, "description": "Demanda inexistente."},
        422: {"model": ErrorResponse, "description": "Identificador invalido."},
    },
)
def get_demand(
    demand_id: UUID,
    session: Annotated[Session, Depends(get_session)],
) -> DemandDetail:
    return DemandDetail.model_validate(DemandService(session).get(demand_id))


@router.patch(
    "/{demand_id}",
    response_model=DemandDetail,
    summary="Editar contexto da demanda",
    description=(
        "Altera somente os campos enviados. description e owner_id enviados como null sao limpos; "
        "etapa, situacao, cliente e artefatos permanecem inalterados."
    ),
    responses={
        404: {"model": ErrorResponse, "description": "Demanda ou responsavel inexistente."},
        422: {"model": ErrorResponse, "description": "Dados de entrada invalidos."},
    },
)
def update_demand(
    demand_id: UUID,
    data: DemandUpdate,
    session: Annotated[Session, Depends(get_session)],
) -> DemandDetail:
    return DemandDetail.model_validate(DemandService(session).update(demand_id, data))

File: web/src/services/demands.ts
/**
 * Chamadas ao gateway para este dominio.
 *
 * Cada funcao corresponde a uma operacao declarada no contrato OpenAPI
 * (packages/contracts/openapi/). Tipos de entrada e saida vivem em `src/types`
 * e derivam do mesmo contrato (RNF02).
 *
 * Estado do contrato: `createDemand`, `getDemand` e `updateDemand` ja estao
 * declaradas em `pipeline-service.yaml`, e entrada e saida espelham os schemas
 * de la. Atencao ao retorno das duas ultimas: o contrato devolve `DemandDetail`
 * - a demanda com o cliente e os artefatos aninhados -, e nao `DemandRead`,
 * como este modulo prometia enquanto a operacao nao estava publicada.
 */

import { api, type RequestContext } from '@/services/api';
import type {
  DemandCreate,
  DemandDetail,
  DemandListParams,
  DemandRead,
  DemandUpdate,
  Page,
  Uuid,
} from '@/types/api';

const RESOURCE = '/demands';

/**
 * Cria uma demanda vinculada a um cliente (RF02).
 *
 * Situacao (`ABERTA`) e etapa inicial (`CAPTACAO`) sao definidas pelo servico:
 * o corpo nao as envia.
 */
export function createDemand(
  payload: DemandCreate,
  context: RequestContext = {},
): Promise<DemandRead> {
  return api.post<DemandRead, DemandCreate>(RESOURCE, payload, context);
}

/**
 * Lista demandas filtrando por situacao, cliente, etapa, responsavel e periodo
 * (RF03).
 *
 * A paginacao vai em `limit`/`offset`, como o contrato declara - a resposta e
 * que volta em `page`/`size`. Enquanto a operacao nao estava publicada, este
 * modulo enviava `page`/`size` tambem na entrada, e o servidor ignorava.
 */
export function listDemands(
  params: DemandListParams = {},
  context: RequestContext = {},
): Promise<Page<DemandRead>> {
  return api.get<Page<DemandRead>>(RESOURCE, {
    ...context,
    query: {
      limit: params.limit,
      offset: params.offset,
      status: params.status,
      client_id: params.client_id,
      stage: params.stage,
      owner_id: params.owner_id,
      from: params.from,
      to: params.to,
    },
  });
}

/**
 * Busca o contexto de uma demanda pelo identificador (RF02, RF04).
 *
 * A resposta e o agregado `DemandDetail`: alem dos campos da demanda, traz o
 * cliente e os artefatos com suas versoes, de proposito, para a tela de
 * detalhe nao precisar de uma segunda consulta.
 */
export function getDemand(demandId: Uuid, context: RequestContext = {}): Promise<DemandDetail> {
  return api.get<DemandDetail>(`${RESOURCE}/${demandId}`, context);
}

/**
 * Edita titulo, contexto ou responsavel da demanda (RF02).
 *
 * Altera somente os campos enviados; `description` e `owner_id` enviados como
 * `null` sao limpos. Etapa e situacao nao passam por aqui: mudam pelas
 * operacoes de pipeline (RF05, RF06), que registram a transicao no historico.
 *
 * Devolve o mesmo agregado de `getDemand`, ja atualizado - por isso a tela
 * pode escrever a resposta direto no cache em vez de recarregar.
 */
export function updateDemand(
  demandId: Uuid,
  payload: DemandUpdate,
  context: RequestContext = {},
): Promise<DemandDetail> {
  return api.patch<DemandDetail, DemandUpdate>(`${RESOURCE}/${demandId}`, payload, context);
}
